modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTable.java

```diff
@@ -0,0 +1,203 @@
+ */
+public abstract class CacheTable extends InMemoryTable {
+    private int maxSize;
+    boolean cacheExpiryEnabled;
+    SiddhiAppContext siddhiAppContext;
+
+    public void initCacheTable(TableDefinition cacheTableDefinition, ConfigReader configReader,
+                               SiddhiAppContext siddhiAppContext, RecordTableHandler recordTableHandler,
+                               boolean cacheExpiryEnabled, int maxSize, String cachePolicy) {
+        this.maxSize = maxSize;
+        this.cacheExpiryEnabled = cacheExpiryEnabled;
+        this.siddhiAppContext = siddhiAppContext;
+        addRequiredFieldsToCacheTableDefinition(cacheTableDefinition, cacheExpiryEnabled);
+
+        // initialize cache table
+        MetaStreamEvent cacheTableMetaStreamEvent = new MetaStreamEvent();
+        cacheTableMetaStreamEvent.addInputDefinition(cacheTableDefinition);
+        for (Attribute attribute : cacheTableDefinition.getAttributeList()) {
+            cacheTableMetaStreamEvent.addOutputData(attribute);
+        }
+
+        StreamEventFactory cacheTableStreamEventFactory = new StreamEventFactory(cacheTableMetaStreamEvent);
+        StreamEventCloner cacheTableStreamEventCloner = new StreamEventCloner(cacheTableMetaStreamEvent,
+                cacheTableStreamEventFactory);
+        super.initTable(cacheTableDefinition, cacheTableStreamEventFactory,
+                cacheTableStreamEventCloner, configReader, siddhiAppContext, recordTableHandler);
+    }
+
+    public void addStreamEventUptoMaxSize(StreamEvent streamEvent) {
+        int sizeAfterAdding = 0;
+        ComplexEventChunk<StreamEvent> addEventsLimitCopy = new ComplexEventChunk<>(true);
+        StreamEvent streamEventIter;
+        while (true) {
+            sizeAfterAdding++;
+            streamEventIter = streamEvent.getNext();
+            if (sizeAfterAdding == maxSize || streamEventIter.getNext() == null) {
+                streamEventIter.setNext(null);
+                break;
+            }
+        }
+        addEventsLimitCopy.add(streamEvent);
+        readWriteLock.writeLock().lock();
+        try {
+            this.add(addEventsLimitCopy);
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    public void addAndTrimUptoMaxSize(ComplexEventChunk<StreamEvent> addingEventChunk) {
+        ComplexEventChunk<StreamEvent> addingEventChunkForCache = new ComplexEventChunk<>(true);
+        addingEventChunk.reset();
+        while (addingEventChunk.hasNext()) {
+            StreamEvent event = addingEventChunk.next();
+            addingEventChunkForCache.add((StreamEvent) generateEventWithRequiredFields(event, siddhiAppContext,
+                    cacheExpiryEnabled));
+        }
+        readWriteLock.writeLock().lock();
+        try {
+            super.add(addingEventChunkForCache);
+            while (this.size() > maxSize) {
+                this.deleteOneEntryUsingCachePolicy();
```

Comment:
Is it worth improving this like this.deleteEntriesUsingCachePolicy(this.size() - maxSize)?

Reply:
yes it is. I did this for convenience and will add another function